docs/source/format/Columnar.rst

```diff
@@ -0,0 +1,1266 @@
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. _format_columnar:
+
+*********************
+Arrow Columnar Format
+*********************
+
+The "Arrow Columnar Format" includes a language-agnostic in-memory
+data structure specification, metadata serialization, and a protocol
+for serialization and generic data transport.
+
+This document is intended to provide adequate detail to create a new
+implementation of the columnar format without the aid of an existing
+implementation. We utilize Google's `Flatbuffers`_ project for
+metadata serialization, so it will be necessary to refer to the
+project's `Flatbuffers protocol definition files <FlatbuffersFiles>`_
+while reading this document.
+
+The columnar format has some key features:
+
+* Data adjacency for sequential access (scans)
+* O(1) (constant-time) random access
+* SIMD and vectorization-friendly
+* Relocatable without "pointer swizzling", allowing for true zero-copy
+  access in shared memory
+
+Some things to keep in mind:
+
+* Some data types can be mutated in-place in memory, but others
+  cannot. The community has made deliberate trade-offs regarding
+  in-place mutability to provide analytical performance and data
+  locality guarantees
+
+Terminology
+===========
+
+Since different projects have used different words to describe various
+concepts, here is a small glossary to help disambiguate.
+
+* **Array** or **Vector**: a sequence of values with known length all
+  having the same type. These terms are used interchangeably in
+  different Arrow implementations, but we use "array" in this
+  document.
+* **Slot**: a single logical value in an array of some particular data type
+* **Buffer** or **Contiguous memory region**: a sequential virtual
+  address space with a given length. Any byte can be reached via a
+  single pointer offset less than the region's length.
+* **Physical Layout type**: The underlying memory layout for an array
+  without taking into account any value semantics. For example, a
+  32-bit signed integer array and 32-bit floating point array have the
+  same layout type.
+* **Logical type**: An application-facing semantic value type that is
+  implemented using some physical layout type. For example, Decimal
+  values are stored as 16 bytes in a fixed-size binary
+  layout. Similarly, strings can be stored as ``List<1-byte>``. A
+  timestamp may be stored as 64-bit fixed-size layout.
+* **Nested** or **parametric type**: a data type whose full structure
```

Comment:
Hmm, some parametric types (such as timestamps) are not nested. I don't think the parametric nature of a type is very important for understanding, so perhaps drop it here?